### DISCUSSION TITLE
Add official support for React extensions

### DISCUSSION THREAD

**Discussion by eliashussary** [directus/directus#4791] [Feature Requests]
Would the core team here entertain the idea of supporting React as an official framework for writing display/interface/layout extensions?

Happy to run a POC and contribute on this.

**Comment by rijkvanzanten** [comment]
Very interesting! It more or less means that we'll end up writing a compatibility layer that allows you to run a React component within a Vue app 👀 Wondering if such a thing is (easily) doable

**Reply by eliashussary** [reply]
Yeah, the approach wouldn't be exclusive to React.  It would be done with two(?) tiny utils;

- vue "escape hatch"
  - create a ref and pass the element ref to a function along with app context
  - if someone wants to use a more imperative API at that point, that's on them... they can use whatever framework they'd like

- react bindings?? (may not be necessary)
  - thin higher-order-component to handle lifecycle changes as props coming in from vue change 




It'll make Directus more approachable to teams whom already are heavily invested in React and have some components they may want to reuse. For example, I have quite a few Mapbox components that I want my users to interact with for data entry... but i'm not willing to rewrite into Vue :)

**Reply by eliashussary** [reply]
Alright... here is a naive proof of concept

https://github.com/eliashussary/directus-react-extensions-poc/tree/main/extensions/interfaces/my-interface/src

this is really the only thing that's relevant: https://github.com/eliashussary/directus-react-extensions-poc/blob/main/extensions/interfaces/my-interface/src/withDirectusReact.js


Here it is in action

![image](https://user-images.githubusercontent.com/7349258/115166008-f617a400-a07e-11eb-8fb4-5d033302ea03.png)

**Reply by rijkvanzanten** [reply]
This is fascinating! We'll have to do some more research into the performance downsides (if any) of this approach, and also consider if we should bundle react into the core codebase for use in extensions (so extensions don't have to bundle their own copy of react every time)

**Reply by eliashussary** [reply]
I probably wouldn't bundle it into your core, I see it as unnecessary overhead.

You can add a way for users to provide a global, and have these extensions bundled assuming that react is provided is a global provided on `window`.


Another thing you can do is provide a single entrypoint for extensions. - this may actually make some dev tooling easier. I would prefer it this way.

Single entry points will ensure that we're importing an external dep once for all extensions. - this problem is obviously not exclusive to React.

```ts
// ./extensions/index.ts

export default {
   interfaces: [],
   displays: [],
   layouts:[],
   //...
}
```
In a dev environment, you can probably get some nice features out of this like `hmr`.

**Reply by rijkvanzanten** [reply]
> Another thing you can do is provide a single entrypoint for extensions. - this may actually make some dev tooling easier. I would prefer it this way.

This would work for locally built extensions, but won't work with pre-made third party extensions loaded in from external sources 🤔

**Reply by eliashussary** [reply]
Wouldn't pre-made extensions still require it to be installed in a dev environment? - Are you allowing them to be installed from the app at runtime?

**Reply by rijkvanzanten** [reply]
All of these extensions are installed in the app in runtime! Nothing requires rebuilding the app/api

**Reply by eliashussary** [reply]
When you say "installed" do you mean to say that they're required / imported?

I understand the dev process today is `bundle  -> move dist`  - the app/api doesn't do any bundling.

Assuming that extensions _only_ work today by going into the filesystem and dropping the file into an extensions dir, this wouldn't be so different from a user importing an external extension into the `index.ts` file - or some sort of manifest file, it doesn't need to be a `index.ts` file.

```ts
// ./extensions/index.ts
import { ThirdPartyInterface } from "directus-third-party";
import { LocalInterface } from "./foo";
export default {
   interfaces: [ ThirdPartyInterface, LocalInterface ],
   //....
}
```

even if you continue to support the current setup, this would greatly improve development workflows for devs that need to deliver a suite of extensions to the client. 

for example, i could publish a single extension that will provide the user geospatial interfaces, displays, and layouts. - the consumer only needs to install that extension once, and provide it once. oppose to installing each component individually and putting them in their corresponding directories.

if you flatten the extension hierarchy and modify the interface it shouldn't be so difficult

```ts
interface IDirectusExtension {
    id: string;
    type: string;
    namespace: string
    // ... the extension properties 
}

type Extensions = IDirectusExtensions[]
````

we can figure something out for the server too. then a developer who want's to have a simple pluggable extension can have their workspace something along these lines;

```ts
|- client
  |- displays
  |- interfaces
  |-...
|- server
  |- hooks
  |- ...
- package.json
```

the consumer can simply just define the npm module as the plugin and directus will work its magic...

```ts
// .. directus.config.js
{
   plugins: ["@acme/directus-postgis-extensions"]
}
```

you can even handle this at runtime, from a "marketplace" and have these lists of modules in a table of some sort 🤷🏼‍♂️ 

sorry for the long windedness :)

**Reply by rijkvanzanten** [reply]
> Assuming that extensions only work today by going into the filesystem and dropping the file into an extensions dir

This is the big part of the puzzle that's changing very soon 🙂 The plan for extensions is that we pull them dynamically through `npm`, so an extension would be installed by running `npm install directus-extension-endpoint` and each of those can have their own dependencies and dependencies on other directus extensions. The local extensions folder is more meant for "local" / private extensions that don't have to be distributed

> you can even handle this at runtime, from a "marketplace" and have these lists of modules in a table of some sort 

This is the second step after that, we can dynamically pull a listing from NPM based on a pre-defined keyword or name pre/suffix and render that as a plugin-directory within the app

**Reply by rijkvanzanten** [reply]
(cc @nickrum )

**Reply by eliashussary** [reply]
cool, do you have a roadmap for this?

**Reply by rijkvanzanten** [reply]
Hard to give any conclusive timeline, but the distributing to/loading from NPM is on the agenda for somewhere within the next couple weeks!

**Reply by eliashussary** [reply]
do you have any technical details you can share on it?

**Reply by rijkvanzanten** [reply]
There's no formalized (final) spec yet, but the general idea is that we'll use npm programmatically to list all the available modules that have a predefined prefix (something along the lines of `directus-ext-*`). We can then loop over the package.json's for those to extract the type and inject them in Directus like a "regular" extension. 

We're still looking into different ways of sharing dependencies (@Illubyte had some good ideas around that) for front-end extensions, and we gotta make sure this approach of depending on other extensions within extensions works, or if we should add a dedicated "bundle" type that is a single bundled package with multiple extensions contained within. I'd love to hear any and all thoughts you have around this either here or on Discord! 😄

**Reply by eliashussary** [reply]
i'd be happy to contribute to ironing out the spec and being hands-on with reviewing some of development.

my capacity is a bit limited so i can't contribute to hacking on this a bunch, but wherever possible i can :)

this is my discord handle: elias.codes#8842

**Reply by rijkvanzanten** [reply]
@nickrum and I will start an "rfc" we can use as a starting point for more specific discussions 👍🏻

**Comment by AustinPhillipTaylor** [comment]
I would personally love to see React integration for extensions, however, I don't think it's something that would be conducive to the current setup. As it stands, Directus makes available a library of global components that anyone can use in their extensions. Those predefined components are useful in maintaining a consistency of design throughout the app, unfortunately it comes at the cost of requiring the extensions to be aware of the underlying Vue instance. 

From my understanding, there is a plan to eventually extract these global components into their own package, which would be beneficial in cases like this. Though, using another framework would still require those components to be translated into the respective framework.

**Reply by exsurgo** [reply]
That's a good point.   I'm envisioning something [like this example](https://github.com/exsurgo/directus-extension-react-example/blob/master/src/react-example.tsx), where a developer could import the Directus components as React components (likely implemented as thin wrappers), to be used in their React extensions.

**Comment by cj** [comment]
One solution might be to build the components using https://github.com/builderio/jsx-lite

**Comment by raxon** [comment]
Meaby web components?

**Reply by rijkvanzanten** [reply]
Web components should already be compatible to use within Vue 🙂

**Reply by exsurgo** [reply]
Hmm, true.    I can see cases where devs wouldn't want to use VueJS at all for their extensions though.   For instance, maybe they just want to pass in a web component directly.    It might be a good idea to allow the `component` property of modules/layouts/displays/etc to accept an `HTMLElement` (or React Component), in addition to a Vue component.

**Comment by exsurgo** [comment]
I 100% agree that React extensions should be officially supported.   React just has too much traction in the industry.     

On my own project, I'm using the [veaury library](https://github.com/devilwjp/veaury) to interop between React and Vue.   I put together a proof of concept for a Directus module built in React:    [https://github.com/exsurgo/directus-extension-react-example](https://github.com/exsurgo/directus-extension-react-example/blob/master/src/react-example.tsx)

To solve the issue first mentioned by @AustinPhillipTaylor above, I'm wrapping Directus v-components as React components, so they can be pulled into my React extensions easily.    In a perfect world, it would probably be better to implement those Directus components as web components and wrap them with thin Vue/React/etc wrappers, but this would be a significant effort.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
